### PR TITLE
[dask-on-ray] Fix Dask-on-Ray scheduler break caused by changing internal Dask API.

### LINKED ISSUE #15094: [dask-on-ray] Update Dask-on-Ray scheduler to be compatible with concurrent.futures refactor.
Dask [recently refactored](https://github.com/dask/dask/commit/ad0e5d140dfc934d94583f8ce50d6b00a04cea38) their local scheduler to be based on `concurrent.futures`, breaking some our dependence on changed/removed internal Dask APIs. We need to update the Dask-on-Ray scheduler to:

1. [P0] Either (a) be compatible with the `concurrent.futures`-based API, or (b) copy the internal APIs we were relying on previously into the Dask-on-Ray implementation.
2. [P1] Remove the dependence on internal Dask scheduler APIs.

### PR DESCRIPTION
**Pull request by clarkzinzow**
## Why are these changes needed?

Dask recently refactored their local scheduler to be based on `concurrent.futures`, breaking some our dependence on changed/removed internal Dask APIs. This PR fixes that break by copying over the old Dask local scheduler code we were reusing and using that instead.

## Related issue number

Closes #15094

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/dask/scheduler_utils.py [hunk 1]
```diff
@@ -0,0 +1,373 @@
+import os
+from queue import Queue, Empty
+
+from dask import config
+from dask.callbacks import local_callbacks, unpack_callbacks
+from dask.core import (
+    _execute_task, flatten, get_dependencies, has_tasks, reverse_dict)
+from dask.order import order
```
Comment (clarkzinzow) [inline python/ray/util/dask/scheduler_utils.py:8]: Note that we're still relying on internal Dask APIs, but these APIs are much more narrow and much more stable.

**Comment by ericl** [comment]
Can we also bump the pinned Dask version in CI?

**Comment by clarkzinzow** [comment]
@ericl Do you know where that version is pinned? AFAICT, the Dask version isn't pinned: https://github.com/ray-project/ray/blob/6b49714c04937b9ab724ca52b6b05ded788805e0/python/requirements.txt#L52

**Comment by rkooo567** [comment]
Btw while I agree with pinning the version for CI stability, how will we detect the regression from the master Dask?

**Review by rkooo567 — APPROVED** [review]

**Comment by ericl** [comment]
LINT issues

**Comment by ericl** [comment]
Dependabot can make PRs to upgrade automatically, e.g., https://github.com/ray-project/ray/pull/14964

Seems we can add more modules to its config: https://github.com/ray-project/ray/blob/master/.github/dependabot.yml

**Comment by ericl** [comment]
Windows tests is failing only on
//python/ray/tests:test_actor_resources                                 TIMEOUT in 3 out of 3 in 300.3s
  Stats over 3 runs: max = 300.3s, min = 64.1s, avg = 221.6s, dev = 111.3s
  C:/users/runneradmin/_bazel_runneradmin/vlncal46/execroot/com_github_ray_project_ray/bazel-out/x64_windows-opt/testlogs/python/ray/tests/test_actor_resources/test.log


which is flaky on master

**Comment by rkooo567** [comment]
Hey @amogkam This needs to be cherry-picked. I assume it wasn't right?

**Comment by amogkam** [comment]
Regarding the required re-testing- "Dask Unit tests with last 3 Dask versions".

Are multiple Dask versions being tested against in our CI?

**Comment by rkooo567** [comment]
No. Currently we are only running one version. (I assume they are fixed). @clarkzinzow we should probably make a release test to run dask on ray against 3~4 different previous versions.

**Comment by amogkam** [comment]
OK @rkooo567 @clarkzinzow can one of you take care of the testing against the last 3 versions? The commit has been picked already on the release branch.

**Comment by rkooo567** [comment]
I can run it later today.

**Pull request merged**